src/primitives/common/linear_equation.rs

```diff
@@ -3,62 +3,194 @@ use crate::{
+const NORMAL_VECTOR_SCALE: i32 = 1 << 10;
+
+/// Linear equation.
 ///
-/// The equation is stored as the a, b and c coefficients of the ax + by + c = 0 equation
+/// The equation is stored as a normal vector and the distance to the origin.
 #[derive(Copy, Clone, PartialEq, PartialOrd, Debug)]
-pub struct LinearEquation<T> {
-    pub a: T,
-    pub b: T,
-    pub c: T,
+pub struct LinearEquation {
+    /// Normal vector, perpendicular to the line.
+    ///
+    /// The unit vector is scaled up to increase the resolution.
+    pub normal_vector: Point,
+
+    /// Distance from the origin.
+    ///
+    /// The distance doesn't directly correlate to the distance in pixels, but is
+    /// scaled up by the length of the normal vector.
+    pub origin_distance: i32,
 }
 
-impl LinearEquation<i32> {
-    pub const fn from_line(line: &Line) -> Self {
+impl LinearEquation {
+    /// Creates a new linear equation from a line.
+    pub fn from_line(line: &Line) -> Self {
+        let normal_vector = line.delta().rotate_90();
+        let origin_distance = line.start.dot_product(normal_vector);
+
         Self {
-            a: line.end.y - line.start.y,
-            b: line.start.x - line.end.x,
-            c: line.end.x * line.start.y - line.start.x * line.end.y,
+            normal_vector,
+            origin_distance,
         }
     }
-}
 
-impl LinearEquation<Real> {
-    /// Create a new linear equation based on one point and one angle
+    /// Creates a new linear equation based on one point and one angle.
     pub fn from_point_angle(point: Point, angle: Angle) -> Self {
         // FIXME: angle.tan() for 180.0 degrees isn't exactly 0 which causes problems when drawing
         //        a single quadrant. Is there a better solution to fix this?
-        let (a, b) = if angle == Angle::from_degrees(180.0) {
-            (Real::from(0.0), Real::from(-1.0))
+        let normal_vector = if angle == Angle::from_degrees(180.0) {
+            Point::new(0, NORMAL_VECTOR_SCALE)
         } else {
-            match angle.tan() {
-                None => (Real::from(1.0), Real::from(0.0)),
-                Some(a) => (-a, Real::from(-1.0)),
-            }
+            -Point::new(
+                i32::from(angle.sin() * Real::from(NORMAL_VECTOR_SCALE)),
+                i32::from(angle.cos() * Real::from(NORMAL_VECTOR_SCALE)),
+            )
         };
 
-        let c = -(a * point.x.into() + b * point.y.into());
-        LinearEquation { a, b, c }
+        let origin_distance = point.dot_product(normal_vector);
+
+        LinearEquation {
+            normal_vector,
+            origin_distance,
+        }
     }
 
-    /// Create a horizontal line equation
+    /// Creates a horizontal line equation.
     pub fn new_horizontal() -> Self {
         LinearEquation {
-            a: Real::from(0.0),
-            b: Real::from(1.0),
-            c: Real::from(0.0),
+            normal_vector: Point::new(0, -16384),
```

Comment:
Would this make sense to be

```suggestion
            normal_vector: Point::new(0, i32::MIN),
```

I'm not sure what the significance of -16384 is specifically.

Reply:
I did forget to update this. It was supposed to be `-NORMAL_VECTOR_SCALE`, which makes the normal vector have the same length as normal vectors generated by `from_point_angle`.